File: Andy/src/level1/javascript/reportResult.js
//프로그래머스 Level 1,신고 결과 받기
function solution(id_list, report, k) {
    var answer = [];
    var ulist = [];
    var rlist = [];
    var block_list = [];
    
    
    id_list.forEach( (d) => {
        ulist[d] = new Set();
        rlist[d] = new Set();
        block_list[d] = 0;
    });
    
    report.forEach( (d) => {
        let [x,y] = d.split(" ");
        ulist[x].add(y);
        rlist[y].add(x);
    });
       
    for (let d in rlist) {
        if(rlist[d].size >= k)
            block_list[d] = 1;
    }    
    
    id_list.forEach( (d) => {
        let cnt = 0;
        ulist[d].forEach( (d) => {
            if(block_list[d]>0) cnt++;
        });
        answer.push(cnt);
    });
    
    return answer;
}

File: Andy/src/level1/java/reportResult.java
package level1.java;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

//프로그래머스 Level 1,신고 결과 받기
public class reportResult {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        int[] block_list = new int[id_list.length];
        int[] ucnt = new int[id_list.length];
        int[] idx_list = new int[id_list.length];
        
        // report의 중복값 제거
        Set<String> report2 = new HashSet<>();
        for(String d:report) {
        	report2.add(d);
        }
        
        // 빠른 검색을 위하여 id_list정렬
        String[] id_list2 = Arrays.copyOf(id_list, id_list.length);
        Arrays.sort(id_list2);
        
        // 정렬된 유저와 정렬안된 유저를 매칭하기 위한 인덱스List
        for(int i=0;i<id_list.length;i++) {
        	String d = id_list[i];
        	idx_list[i] = Arrays.binarySearch(id_list2, d);
        }
        
        // 신고당한 유저카운트 (각각의 유저가 몇번 신고당했는지카운트값 저장) 
        for(String d:report2) {
        	String[] tmp = d.split(" ");
        	int index2 = Arrays.binarySearch(id_list2, tmp[1]);
        	ucnt[index2]++;
        }
        
        // 신고한 유저들의 각각의 블럭당한 유저 카운트
        for(String d:report2) {
        	String[] tmp = d.split(" ");
        	int index = Arrays.binarySearch(id_list2, tmp[0]);
        	int index2 = Arrays.binarySearch(id_list2, tmp[1]);
        	if(ucnt[index2]>=k) block_list[index]++;
        }
               
        for(int i=0;i<id_list.length;i++) {
        	answer[i] = block_list[idx_list[i]];
        }               
        
        return answer;
    }
}
